### DISCUSSION TITLE
Type inference error with tuple of literal

### DISCUSSION THREAD

**Discussion by lkct** [microsoft/pylance-release#2929] [Enhancement]
<h3> Code Snippet </h3>

```python
a = [(0,)] + [(1,)]
# a = [('a',)] + [('b',)]   # this is similar 
```

<h3> Actual behavior </h3>

```
Operator "+" not supported for types "list[tuple[Literal[0]]]" and "list[tuple[Literal[1]]]"Pylance(reportGeneralTypeIssues)
```

<h3> Workaround </h3>

Add a `# type: ignore` to the line or use explicit annotation as below (in @erictraut 's reply)

<h3> Suggest behavior </h3>

No error at least. It would be better if type inference could give `a: list[tuple[int]]`.

---

EDIT: convert bug report style to discussion.

**Comment by erictraut** [comment]
This is the intended behavior in pyright, the type checker upon which pyright is built. For details on how pyright infers types, please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#tuple-expressions).

If you want this to type check without errors, you'll need to introduce a type annotation somewhere, like this:
```python
a: list[tuple[int, ...]] = [(0,)]
a += [(1,)]
```

**Comment by lkct** [comment]
@erictraut Thanks for the response. I see that the behaviour is in line with the documentation.

However I still think it should not raise an error. Maybe this could be a feature request instead of a bug?

It's ok with a strict typing on a tuple itself, but intuitively `[(0,)]+[(1,)]` should be similar to `[0]+[1]` which causes no error. If it's difficult to implement an elevation from `tuple[Literal[0|1]]` to `tuple[int]` in list typing, it could at least infer `list[Unknown]`  similar to [item 3 here](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#list-expressions).

**Reply by erictraut** [reply]
Changing the behavior to make this example work would break other common use cases. That's the challenge with type inference. The type inference rules used by pyright have been tuned over multiple years to minimize false positives and reduce the occurrence of unexpected behaviors, but these cannot be eliminated entirely. We're unlikely to change the behaviors at this point because many code bases are now type-checked with pyright and pylance, and changes will generate new false positives in those code bases.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
